In [4]:
import numpy as np

app = 'Anthony Constant Joshua Fatima Julien Bassem Caroline Dan Ines Nidhal Sacia Xavier Roger Hachem Jean-Pierre Myriam Ludo Olivier Pierre-Etienne Wiem Cecilia'.split()
np.random.seed(1)
rapp = [app[i] for i in np.random.choice(21, 21, replace=False)]
for i in range(5):
    print(rapp[4*i:4*i+4] + ([rapp[-1]] if i == 4 else []))

['Myriam', 'Sacia', 'Fatima', 'Pierre-Etienne']
['Olivier', 'Jean-Pierre', 'Ludo', 'Julien']
['Joshua', 'Caroline', 'Dan', 'Constant']
['Hachem', 'Anthony', 'Cecilia', 'Wiem']
['Nidhal', 'Ines', 'Roger', 'Xavier', 'Bassem']


In [212]:
import nltk
nltk.download()
#the most populars packages

showing info https://raw.githubusercontent.com/nltk/nltk_data/gh-pages/index.xml


True

In [213]:
from nltk.corpus import stopwords
print(set(stopwords.words('english')))

{'and', "wouldn't", 'over', 'yourselves', 'off', 'why', 'couldn', "it's", 'that', 'wouldn', 'where', "you'll", 'so', 'very', 'when', "wasn't", 'both', 'such', "aren't", 'mightn', "won't", 'have', 'be', 'him', 'until', 'an', "should've", 'theirs', 'how', 'having', 'nor', 'about', 'his', 'their', 'themselves', 'were', "weren't", 'hasn', 'while', 'being', 'ourselves', 'himself', 're', 'a', 'than', 'them', 'was', 'weren', 'whom', 'did', 'doesn', "hadn't", 'but', 'same', 'after', 'didn', 'or', 'between', "didn't", 'there', 'above', 'all', 'on', 'isn', 'mustn', 'below', 't', "that'll", 'few', 'against', 'not', "you'd", 'as', 'does', 'own', 'wasn', 'am', 'of', "hasn't", "haven't", 'at', "you've", 'll', "she's", "mightn't", 'yourself', "couldn't", 'myself', 'which', 'aren', 'has', 'too', 'your', 'are', 'no', 'our', 'for', 've', 'ma', 'itself', 'd', 'more', 'haven', "needn't", 'she', 'here', 'ours', 'i', 'under', 'y', 'do', 'you', 'shan', 'because', 'will', "shan't", 'her', "mustn't", 'some', '

In [247]:
# récupérer les donées : 1ere méthode

f = open('SMSSpamCollection.txt', 'r') 
data = f.read()

print(f.readline())

In [18]:

from nltk.corpus import stopwords  
from nltk.tokenize import word_tokenize  
  
example_sent = """This is a sample sentence, 
                  showing off the stop words filtration."""
  
stop_words = set(stopwords.words('english'))  
  
word_tokens = word_tokenize(example_sent)  
  
filtered_sentence = [w for w in word_tokens if not w in stop_words]  
  
filtered_sentence = []  
  
for w in word_tokens:  
    if w not in stop_words:  
        filtered_sentence.append(w)  
  
print(word_tokens)  
print(filtered_sentence)  


['This', 'is', 'a', 'sample', 'sentence', ',', 'showing', 'off', 'the', 'stop', 'words', 'filtration', '.']
['This', 'sample', 'sentence', ',', 'showing', 'stop', 'words', 'filtration', '.']


In [319]:
#2 eme méthode pour récupérer les données 

dataset=pd.read_csv("SMSSpamCollection.txt",delimiter="\t", header=None)
dataset

0                                                  1
0      ham  Go until jurong point, crazy.. Available only ...
1      ham                      Ok lar... Joking wif u oni...
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...
3      ham  U dun say so early hor... U c already then say...
4      ham  Nah I don't think he goes to usf, he lives aro...
...    ...                                                ...
5567  spam  This is the 2nd time we have tried 2 contact u...
5568   ham               Will ü b going to esplanade fr home?
5569   ham  Pity, * was in mood for that. So...any other s...
5570   ham  The guy did some bitching but I acted like i'd...
5571   ham                         Rofl. Its true to its name

[5572 rows x 2 columns]

In [249]:
dataset[0]

0        ham
1        ham
2       spam
3        ham
4        ham
        ... 
5567    spam
5568     ham
5569     ham
5570     ham
5571     ham
Name: 0, Length: 5572, dtype: object

In [250]:
dataset[1]

0       Go until jurong point, crazy.. Available only ...
1                           Ok lar... Joking wif u oni...
2       Free entry in 2 a wkly comp to win FA Cup fina...
3       U dun say so early hor... U c already then say...
4       Nah I don't think he goes to usf, he lives aro...
                              ...                        
5567    This is the 2nd time we have tried 2 contact u...
5568                 Will ü b going to esplanade fr home?
5569    Pity, * was in mood for that. So...any other s...
5570    The guy did some bitching but I acted like i'd...
5571                           Rofl. Its true to its name
Name: 1, Length: 5572, dtype: object

In [320]:
# préparation du corpus "features"
corpus = list(dataset[1])


In [321]:
# preparation des targets
targets = list(dataset[0])
targets = [1 if x == "ham" else 0 for x in targets]

In [322]:
from sklearn.feature_extraction.text import CountVectorizer

#vectoriser = countVectorizer() sans stop_words
cv = CountVectorizer()

X = cv.fit_transform(corpus)

In [359]:
# la taille de la matrice obtenue
print(X.shape)
print(X.size)

(5572, 8713)
74169


In [260]:
print(X)

  (0, 3571)	1
  (0, 8084)	1
  (0, 4374)	1
  (0, 5958)	1
  (0, 2338)	1
  (0, 1316)	1
  (0, 5571)	1
  (0, 4114)	1
  (0, 1767)	1
  (0, 3655)	1
  (0, 8548)	1
  (0, 4501)	1
  (0, 1765)	1
  (0, 2061)	1
  (0, 7694)	1
  (0, 3615)	1
  (0, 1082)	1
  (0, 8324)	1
  (1, 5538)	1
  (1, 4537)	1
  (1, 4342)	1
  (1, 8450)	1
  (1, 5567)	1
  (2, 4114)	1
  (2, 3373)	1
  :	:
  (5570, 4245)	1
  (5570, 8371)	1
  (5570, 1097)	1
  (5570, 4642)	1
  (5570, 7089)	1
  (5570, 3323)	1
  (5570, 7674)	1
  (5570, 1451)	1
  (5570, 5367)	1
  (5570, 2606)	1
  (5570, 8120)	1
  (5570, 1794)	1
  (5570, 7099)	1
  (5570, 2905)	1
  (5570, 3489)	1
  (5570, 1802)	1
  (5570, 3709)	1
  (5570, 4188)	1
  (5570, 914)	1
  (5570, 1561)	1
  (5571, 7806)	1
  (5571, 5276)	1
  (5571, 4253)	2
  (5571, 7938)	1
  (5571, 6548)	1


In [351]:
# le nombre de mots se trouvant dans la matrice obtenu 

ListWord = np.argsort(X.sum(0))
ListWord[[0]]

matrix([[4356, 4861, 4860, ..., 7674, 8668, 7806]])

In [352]:
# les 10 premiers mots et derniers mots
#print(cv.get_feature_names()[830:860])
#print(cv.get_feature_names()[-10:-1])
print (ListWord[0])

[[4356 4861 4860 ... 7674 8668 7806]]


In [355]:
# le premier mot qui porte un sens 
print(cv.get_feature_names()[859])

# dans un deuxième temps nous allons procéder à l'élimination des 858 premières rentrées car cela représente que des chiffres.

ListWordWithNbr = cv.get_feature_names()[859:]
print(ListWordWithNbr)

#Nous envisageaons de l'utiliser dans le model d'apprentissage qui donne le meilleur résultat.

abbey
['abbey', 'abdomen', 'abeg', 'abel', 'aberdeen', 'abi', 'ability', 'abiola', 'abj', 'able', 'abnormally', 'about', 'aboutas', 'above', 'abroad', 'absence', 'absolutely', 'absolutly', 'abstract', 'abt', 'abta', 'aburo', 'abuse', 'abusers', 'ac', 'academic', 'acc', 'accent', 'accenture', 'accept', 'access', 'accessible', 'accidant', 'accident', 'accidentally', 'accommodation', 'accommodationvouchers', 'accomodate', 'accomodations', 'accordin', 'accordingly', 'account', 'accounting', 'accounts', 'accumulation', 'achan', 'ache', 'achieve', 'acid', 'acknowledgement', 'acl03530150pm', 'acnt', 'aco', 'across', 'act', 'acted', 'actin', 'acting', 'action', 'activ8', 'activate', 'active', 'activities', 'actor', 'actual', 'actually', 'ad', 'adam', 'add', 'addamsfa', 'added', 'addicted', 'addie', 'adding', 'address', 'adds', 'adewale', 'adi', 'adjustable', 'admin', 'administrator', 'admirer', 'admission', 'admit', 'adore', 'adoring', 'adp', 'adress', 'adrian', 'adrink', 'ads', 'adsense', 'ad

In [353]:
feature_names = cv.get_feature_names()
feature_array = np.array(feature_names)
ListWord = np.argsort(X.sum(0))
ListWord 
# les 10 mots les plus fréquents
n = 10
top_n = feature_array[ListWord][:10]
print(top_n)


[['js' 'maggi' 'magazine' ... 'the' 'you' 'to']]


In [262]:
# Stop words avec NLTK
stop_words = nltk.corpus.stopwords.words("english")
#vectoriser = countVectorizer() avec stop_words
cv = CountVectorizer(stop_words = stop_words)
X = cv.fit_transform(corpus)

In [366]:
X1 = cv.fit_transform(ListWordWithNbr)
print("shape X1 with out number: ", X1.shape)
print("size X1: ",X1.size)
print("Nous remarquons la différence de taille entre les deux matrices")
print("shape X with number : ", X.shape)
print("size X: ",X.size)


shape X1 with out number:  (7854, 7854)
size X1:  7854
Nous remarquons la différence de taille entre les deux matrices
shape X with number :  (5572, 8713)
size X:  74169


In [263]:
# le nombre de mots à éliminer
len(stop_words)

179

In [144]:
# model logistic regression


from sklearn.linear_model import LogisticRegression


from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score

from sklearn import metrics
from  sklearn.Stop words avec NLTKStop words avec NLTKmodel_selection import train_test_split
from sklearn.metrics import classification_report


from time import time


In [229]:

model = LogisticRegression(random_state=42)
model.fit(X,targets)
S = model.score(X,targets)

print("Train score:", S )



Train score: 0.9964106245513281


In [164]:
from  sklearn.model_selection import train_test_split



In [141]:
X_train.shape


(3900, 8577)

###  3) Apprenstissage

####  LogisticRegression

In [295]:
# jeu test1 : test_size =0.2
X_train, X_test, y_train, y_test = train_test_split(X,targets, test_size =0.2, random_state=42)

model = LogisticRegression(random_state=42, max_iter = 200)

start = time()
model.fit(X_train,y_train)
T1= time()-start

y_pred = model.predict(X_test)

print("Time :", T1)
print ("================")
print(classification_report(y_test, y_pred))

print("=======")
conf = confusion_matrix(y_test, y_pred)
conf



Time : 0.1344740390777588
              precision    recall  f1-score   support

           0       1.00      0.91      0.95       149
           1       0.99      1.00      0.99       966

    accuracy                           0.99      1115
   macro avg       0.99      0.96      0.97      1115
weighted avg       0.99      0.99      0.99      1115



array([[136,  13],
       [  0, 966]])

In [314]:
model.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 3,
 'p': 2,
 'weights': 'uniform'}

In [296]:
# jeux de test2 : test_size =0.3
X_train, X_test, y_train, y_test = train_test_split(X,targets, test_size =0.3, random_state=42)

model = LogisticRegression(random_state=42, max_iter = 200)

start = time()
model.fit(X_train,y_train)
T2= time()-start



y_pred = model.predict(X_test)

print("Time :", T2)
print ("================")
print(classification_report(y_test, y_pred))

print("=======")
conf = confusion_matrix(y_test, y_pred)
conf

Time : 0.12093234062194824
              precision    recall  f1-score   support

           0       0.99      0.90      0.94       224
           1       0.99      1.00      0.99      1448

    accuracy                           0.99      1672
   macro avg       0.99      0.95      0.97      1672
weighted avg       0.99      0.99      0.98      1672



array([[ 202,   22],
       [   3, 1445]])

In [309]:
# jeux de test2 : test_size =0.5
X_train, X_test, y_train, y_test = train_test_split(X,targets, test_size =0.5, random_state=42)

model = LogisticRegression(random_state=42, max_iter = 200)

start = time()
model.fit(X_train,y_train)
T3= time()-start



y_pred = model.predict(X_test)

print("Time :", T3)
print ("================")
print(classification_report(y_test, y_pred))


print("=======")
conf = confusion_matrix(y_test, y_pred)
conf


Time : 0.11784863471984863
              precision    recall  f1-score   support

           0       0.98      0.90      0.94       370
           1       0.98      1.00      0.99      2416

    accuracy                           0.98      2786
   macro avg       0.98      0.95      0.96      2786
weighted avg       0.98      0.98      0.98      2786



array([[ 332,   38],
       [   7, 2409]])

In [306]:
# jeux de test2 : test_size =0.8
X_train, X_test, y_train, y_test = train_test_split(X,targets, test_size =0.8, random_state=42)

model = LogisticRegression(random_state=42, max_iter = 200)

start = time()
model.fit(X_train,y_train)
T4= time()-start



y_pred = model.predict(X_test)

print("Time :", T4)
print ("================")
print(classification_report(y_test, y_pred))


print("=======")
conf = confusion_matrix(y_test, y_pred)
conf


Time : 0.17972207069396973
              precision    recall  f1-score   support

           0       0.98      0.83      0.90       593
           1       0.97      1.00      0.99      3865

    accuracy                           0.97      4458
   macro avg       0.98      0.91      0.94      4458
weighted avg       0.97      0.97      0.97      4458



array([[ 490,  103],
       [  11, 3854]])

Le meilleur résultat obtenu avec la regréssion logistique est celui du jeu de test1 : test_size =0.2.
Il permet d'obtenir les résultats suivants :

  precision    recall  f1-score   support

           0       1.00      0.91      0.95       149
           1       0.99      1.00      0.99       966
           
           
 Et une matrice de confusion quin présente les meilleures présication 
 
 array([[136,  13],
       [  0, 966]])
       


#### Nearest Neighbors Classification

In [313]:
#jeu test5 : test_size =0.5
X_train, X_test, y_train, y_test = train_test_split(X,targets, test_size =0.5, random_state=42)


from sklearn.neighbors import KNeighborsClassifier 
model = KNeighborsClassifier(n_neighbors=3) 
start = time()
model.fit(X_train,y_train)
T5= time()-start

 
Y_pred = model.predict(X_test)


print("Time :", T5)
print ("================")
print(classification_report(y_test, y_pred))



conf = confusion_matrix(y_test, y_pred)
conf

S = model.score(X_train,y_train)
S1 = round(model.score(X_test,y_test),4)
print ('By score() :')
print("Accuracy train  :", S )
print ("test_size =0.5")
print("Accuracy test :", S1 ) 
 
print("=======")    



Time : 0.0011990070343017578
              precision    recall  f1-score   support

           0       0.98      0.90      0.94       370
           1       0.98      1.00      0.99      2416

    accuracy                           0.98      2786
   macro avg       0.98      0.95      0.96      2786
weighted avg       0.98      0.98      0.98      2786

By score() :
Accuracy train  : 0.9404163675520459
test_size =0.5
Accuracy test : 0.9203


In [315]:
model.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 3,
 'p': 2,
 'weights': 'uniform'}

#### Decision Trees

In [311]:
# jeu test6 : test_size =0.5

from sklearn import tree
model = tree.DecisionTreeClassifier()

start = time()
model.fit(X_train,y_train)
T6= time()-start

Y_pred = model.predict(X_test)

print("Time :", T5)
print ("================")
print(classification_report(y_test, y_pred))


print("=======")
conf = confusion_matrix(y_test, y_pred)
conf



Time : 0.0019736289978027344
              precision    recall  f1-score   support

           0       0.98      0.90      0.94       370
           1       0.98      1.00      0.99      2416

    accuracy                           0.98      2786
   macro avg       0.98      0.95      0.96      2786
weighted avg       0.98      0.98      0.98      2786



array([[ 332,   38],
       [   7, 2409]])

In [316]:
model.get_params()

{'algorithm': 'auto',
 'leaf_size': 30,
 'metric': 'minkowski',
 'metric_params': None,
 'n_jobs': None,
 'n_neighbors': 3,
 'p': 2,
 'weights': 'uniform'}

Avec le test size = 0.5 la regression logistique et l'arbre de décision donnent les mêmes résultats. Cependant, l'arbre de décison est plus rapide avec  avec un temps de  0.00197au lieu de  0.11784 pour la regression linéaire.
Nous remarquons , également, que la commande model.get_params() donne le même résultat pour les 3 models testés).

In [367]:
# jeu test1.1: test_size =0.2 et X 1 with out numbers
X_train, X_test, y_train, y_test = train_test_split(X1,targets, test_size =0.2, random_state=42)

model = LogisticRegression(random_state=42, max_iter = 200)

start = time()
model.fit(X_train,y_train)
T11= time()-start

y_pred = model.predict(X_test)

print("Time :", T11)
print ("================")
print(classification_report(y_test, y_pred))

print("=======")
conf = confusion_matrix(y_test, y_pred)
conf



ValueError: Found input variables with inconsistent numbers of samples: [7854, 5572]

In [ ]:
# cross valisation
from sklearn.model_selection import ShuffleSplit

In [230]:
cv = ShuffleSplit(n_splits=5, test_size=0.20, random_state=42)

In [239]:
from sklearn.model_selection import cross_val_score
start = time()
lr = model.fit(X,targets)
scores = cross_val_score(lr, X, targets, cv=5)
T2= time()-start

In [241]:
print(scores)
print("time : ", T2)

[0.9838565  0.98206278 0.98114901 0.97755835 0.98204668]
time :  0.6234207153320312
